lms/security.py

```diff
@@ -143,19 +143,29 @@ def _get_lti_user(request):
         partial(bearer_token_schema.lti_user, location="form"),
         OAuthCallbackSchema(request).lti_user,
     ]
-    # Avoid checking for the LTI1.3 JWT if not there
-    if "id_token" in request.params:
-        # Don't replace all auth methods in case somehow we have a bogus token
-        # but try this one first.
-        schemas.insert(0, LTI13AuthSchema(request).lti_user)
 
     lti_user = None
-    for schema in schemas:
+    # Avoid checking for the LTI1.3 JWT if not there
+    if "id_token" in request.params:
         try:
-            lti_user = schema()
-            break
-        except ValidationError:
-            continue
+            lti_user = LTI13AuthSchema(request).lti_user()
+        except ValidationError as err:
+            from lms.models import LTIUser  # pylint:disable=import-outside-toplevel
+
+            # If we haven't been able to get an LTIUser still return
+            # an empty one including the ValidationError.
+            # We'll be able to use the exception later in the exception view.
+            #
+            # Directly raising here won't be picked by the exception views.
```

Comment:
> Directly raising here won't be picked by the exception views

I can't point to the right place in the docs for this, maybe here: https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/router.html

I reckon that makes sense, here we are worry about security, not building a response.

It's our design on top, trying many methods at once, which make giving feedback tricky.

Reply:
Yep, it doesn't surprise me that an exception raised by authentication code wouldn't be picked up by the exception views: the exception views catch exceptions raised by _the views_, they don't catch exceptions raised by other parts of the request processing pipeline (like authentication). I don't think authentication code is supposed to raise exceptions up to Pyramid